packages/labs/signals/src/lib/signal-watcher.ts

```diff
@@ -0,0 +1,176 @@
+     * outside of a watch() directive update. It is set to `false` when
+     * update() is called, so that a requestUpdate() is required to do another
+     * full render.
+     */
+    private __doFullRender = true;
+
+    /**
+     * Set of watch directives that have been updated since the last update.
+     * These will be committed in update() to ensure that the latest value is
+     * rendered and that all updates are batched.
+     */
+    private __pendingWatches = new Set<WatchDirective<unknown>>();
+
+    protected override performUpdate() {
+      // Always enable watching before an update, even if disconnected, so that
+      // we can track signals that are accessed during the update.
+      this.__watch();
+      // Force an uncached read of __performUpdateSignal
+      this.__forcingUpdate = true;
+      this.__forceUpdateSignal?.set(this.__forceUpdateSignal.get() + 1);
+      this.__forcingUpdate = false;
+      // Always read from the signal to ensure that it's tracked
+      this.__performUpdateSignal!.get();
+    }
+
+    protected override update(
+      changedProperties: PropertyValueMap<this> | Map<PropertyKey, unknown>
+    ): void {
+      if (this.__doFullRender) {
+        // Force future updates to not perform full renders by default.
+        this.__doFullRender = false;
+        super.update(changedProperties);
+      } else {
+        // For a partial render, just commit the pending watches.
+        this.__pendingWatches.forEach((d) => d.commmit());
+        this.__pendingWatches.clear();
+        // Since we don't call super.update(), we need to set this to false
+        this.isUpdatePending = false;
+      }
+    }
+
+    override requestUpdate(
+      name?: PropertyKey | undefined,
+      oldValue?: unknown,
+      options?: PropertyDeclaration<unknown, unknown> | undefined
+    ): void {
+      this.__doFullRender = true;
+      super.requestUpdate(name, oldValue, options);
+    }
+
+    override connectedCallback(): void {
+      this.__watch();
+      super.connectedCallback();
+      // Because we might have missed some signal accesses while disconnected,
+      // we need to force a full render on the next update.
+      this.requestUpdate();
+    }
+
+    override disconnectedCallback(): void {
+      super.disconnectedCallback();
+      // Clean up the watcher to avoid memory leaks from signals holding
+      // references to the element. This means that while disconnected, regular
+      // reactive property updates will trigger a re-render, but signal updates
+      // will not. To ensure that current signal usage is still correctly
+      // tracked, we re-enable watching in performUpdate() even while
+      // disconnected. From that point on, a disconnected element will be
+      // retained by the signals it accesses during the update lifecycle.
+      // We may want to use a WeakMap inside the watcher to avoid a strong
+      // reference to the element.
```

Comment:
@sorvell whether or not to disable watching while disconnected and re-enable it when updates trigger while disconnected is a big thing to decide here.

I don't love that if we disable watching that we diverge from how reactive updates work while disconnected. Maybe a WeakRef strategy is how we keep updates enabled while avoid the memory leak.

Reply:
This is definitely pretty tricky. Aside from the memory issue there's also just the problem of indeterminate unbounded work. This typically only happens with shared state, and we're otherwise relatively careful to unhook any sharing in disconnected (e.g. [context](https://github.com/lit/lit/blob/main/packages/context/src/lib/controllers/context-consumer.ts#L87)).